rfcs/0021-comparison-and-aggregate-meanings.md

```diff
@@ -0,0 +1,82 @@
+
+## Proposal
+
+### New `ComparisonOperatorDefinition` types
+
+Add the following variants to `ComparisonOperatorDefinition`:
+
+```rust
+ pub enum ComparisonOperatorDefinition {
+     // ...
+     LessThan,
+     LessThanOrEqual,
+     GreaterThan,
+     GreaterThanOrEqual,
+     // ...
+ }
+```
+
+Each of these needs to be given well-defined semantics in the specification.
+
+There are two ways we can go about extracting a specification here: either based on the likely capabilities of our various connectors, or based on the needs of our clients.
+
+If we go the first route, we're likely to run into a variety of unusual implementations of these operators, e.g. databases will depend on collation settings, some operators may not be intuitive, e.g. operators may not be transitive, may use 3-valued logic, etc.
+
+For certain type representations, we could use existing specs, e.g. IEEE 754 defines these operators for floating point numbers. However, for other types, it will depend on the database in question, e.g. collation settings will affect these operators for strings.
+
+If we start with the needs of the clients then what does SQL/datafusion require of these operators?
+
+- A quick look through the datafusion code shows that it rewrites `NOT (a < b)` into `a >= b`, and `a < b` into `b > a` in some cases for canonicalization. We should require these rules on the NDC side.
+- datafusion also implements an interval arithmetic and constraints library, so we should take a look to see what that implies for these operators too.
+- In general, our constraints may change in future if new rewrites get added, so we would need to be careful there.
+
+### Add meanings for aggregate functions
+
+Change `AggregateFunctionDefinition` into an `enum`:
+
+```rust
+pub enum AggregateFunctionDefinition {
+    Sum {
+        result_type: Type,
+    },
+    Min,
+    Max,
+    Custom {
+        result_type: Type,
+    },
+}
+```
+
+Again, these will need to be standardized.
+
+We can require `Sum` to be only defined on types with numeric type representations, and then simply define them in terms of IEEE 754 floats or integers.
+
+`Min` and `Max` ought to also be definable on other types like strings and date/times.
+
+We can require these aggregates to be commutative and monoidal.
+
+## Future Work
+
+Other possible comparison operators (from https://datafusion.apache.org/user-guide/sql/operators.html#other-operators):
+
+- `IS DISTINCT FROM`, `IS NOT DISTINCT FROM`
+- `~`, `~*`, `!~`, `!~*` (regex-based matches, may be difficult to standardize across implementions, might consider "starts with", "contains" and "ends with" instead)
+- `@>`, `<@` (array operators)
+
+Other possible aggregate functions (from https://datafusion.apache.org/user-guide/sql/operators.html#comparison-operators):
```

Comment:
Prior art on standardized aggregate functions: https://substrait.io/extensions/functions_arithmetic/#aggregate-functions.

Reply:
If we add more aggregate functions to NDC, would that be a breaking change?

Reply:
> If we add more aggregate functions to NDC, would that be a breaking change?

It would be breaking for engine since it would need to handle more cases. But if engine is checking the NDC version to make sure it's compatible with its own max-supported-version (not sure if we do this), then it's not a concern.